Ed-Fi-Exchange-OSS/EdFi-Project-Buzz
# [FIF 183] Restrict staff to adding survey answers to only their students.

When added, this commit will restrict teachers to upload survey answers for only their students. Tools for creating sample survey data files are included.

### TESTING
> NOTE: This assumes you are running a database with Grandbend data. The sample SurveyData SQL files can used to create survey files specific for your dataset.
- Be sure to have staff who can upload surveys that matches your survey student data, and a second teacher for throwing errors when you use the same survey files. 1037 cannot load surveys for 1030's students.
```sql
UPDATE buzz.staff set isadminsurveyloader = true where staffkey = 1030; -- for a staff that should work with no errors.
UPDATE buzz.staff set isadminsurveyloader = true where staffkey = 1037; -- a teacher running 1030's files should generate errors

```
-- You can then run the sampleSurveyData SQL files to create student records for that staffkey.
- Change to the ETL folder.
- Run graphile-worker for your database...
```powershell
> npx graphile-worker -c "postgresql://postgres:pa55w0rd@localhost/edfi_buzz"
```
- Manually create a survey job by running the following SQL. (edited). NOTE THAT max_attempts is set to only attempt failed jobs twice.

The following SQL assumes that there are two users staffkeys 1030, and 1037. Each has isadminsurveyloader set to true. There are two surveys for the contact and internet access - there are SQL files for generating samples using 1030 as the staffkey in .\surveySampleData. The files are assumed to also be in the surveySampleData folder.

### EXPECTED RESULT
1. The first file should run but throw an exception that the staffkey cannot load the students in the file. 
2. The second file should load successfully. 
3. The file should fail because it does not exist, and the process should continue.
4. The last file should load successfully.

```sql

-- SHOULD NOT LOAD WRONG STUDENTS
SELECT graphile_worker.add_job('buzzSurvey', json_build_object('staffkey', '1037', 'filename', 'staffkey-1030-InternetAccessSurvey.csv', 'title', 'Internet Access Survey', 'path', 'C:\dev\Ed-Fi\EdFi-Project-Buzz\EdFi.Buzz.Etl\surveySampleData'), max_attempts := 2);

-- SHOULD LOAD CONTACT SURVEY
SELECT graphile_worker.add_job('buzzSurvey', json_build_object('staffkey', '1030', 'filename', 'staffkey-1030-ContactSurvey.csv', 'title', 'Contact Survey', 'path', 'C:\dev\Ed-Fi\EdFi-Project-Buzz\EdFi.Buzz.Etl\surveySampleData'), max_attempts := 2);

-- BAD FILENAME SHOULD NOT CRASH THE WORKER PROCESS AND NEXT FILE SHOULD LOAD
SELECT graphile_worker.add_job('buzzSurvey', json_build_object('staffkey', '1030', 'filename', 'THISFILEDOESNOTEXIST.csv', 'title', 'Contact Survey', 'path', 'C:\dev\Ed-Fi\EdFi-Project-Buzz\EdFi.Buzz.Etl\surveySampleData'), max_attempts := 2);

-- SHOULD LOAD INTERNET ACCESS SURVEY
SELECT graphile_worker.add_job('buzzSurvey', payload:= json_build_object('staffkey', '1030', 'filename', 'staffkey-1030-InternetAccessSurvey.csv', 'title', 'Internet Access', 'path', 'C:\dev\Ed-Fi\EdFi-Project-Buzz\EdFi.Buzz.Etl\surveySampleData'), max_attempts := 2);

```
